1단계 - REST API 및 테스트 리팩토링

testrace opened the pull request:
안녕하세요 리뷰어님 😃 
1단계 리뷰 요청 드립니다!

뼈대 코드가 워낙 잘 되어 있어서 쉽게 봤다가 꽤나 고생했습니다! 생각만큼 쉽지 않더라구요 😢  
스프링의 `@RequestBody` 를 흉내내어 만들까 하다가 PostData가 필요한 곳에만 적용할 수 있도록 구현해 봤습니다.

많은 피드백 부탁드립니다!

testrace committed:
docs: 요구사항 및 기능 목록 정의

testrace committed:
docs: 요구사항 및 기능 목록 정의

testrace committed:
feat: JsonUtils#toJson Java Object를 Json 문자열로 변환한다

testrace committed:
feat: JsonView 응답 헤더 추가

응답 헤더 ContentType을 'application/json'으로 설정한다

testrace committed:
feat: JsonView 응답 데이터 추가

응답 데이터가 1개인 경우

testrace committed:
feat: JsonView 응답 데이터 추가

응답 데이터의 요소가 2개 이상인 경우

testrace committed:
feat: JsonUtils json 문자열에서 특정 파라미터 조회

json 문자열에서 특정 파라미터를 조회한다.
파라미터가 존재하지 않을 경우 null을 리턴한다.

testrace committed:
feat: QueryString or RequestBody에서 파라미터 조회

클라이언트의 요청 파라미터를 QeuryString 혹은 RequestBody에서 구한다
RequestBody는 queryString format과 json format 모두 지원한다

testrace committed:
refactor: ArgumentResolver 구현체에 RequestParameterUtils 적용

testrace committed:
feat: UserApiController 회원 가입

애노테이션 기반 MVC가 적용된 UserApiController 구현
회원 가입 기능 구현

testrace committed:
feat: UserApiController 회원 조회

회원 조회 기능 구현
jackson 라이브러리에서 Json 결과를 자바 빈으로 변환하기 위해
User 클래스에 private 기본 생성자 추가

testrace committed:
fix: RequestParamArgumentResolver 롤백

RequestParamArgumentResolver에서 RequestParameterUtils를 적용 후
HttpServletRequest의 PostData를 읽기 위해 InpuStream을 사용을 했기 때문에
이후 ModelArgumentResolver에서 PostData를 다시 못읽는 현상으로 인해
RequestParameterUtils를 사용하지 않도록 변경함

testrace committed:
feat: UserApiController 회원 수정

회원 수정 기능 구현

yongdae commented on README.md:
```diff
@@ -7,3 +7,24 @@
 ## 온라인 코드 리뷰 과정
 * [텍스트와 이미지로 살펴보는 온라인 코드 리뷰 과정](https://github.com/next-step/nextstep-docs/tree/master/codereview)
+
+# 🚀 1단계 - REST API 및 테스트 리팩토링
+
+### 요구사항
+> HTML 이외에 JSON으로 데이터를 요청하고 응답을 받도록 지원해야 한다.
+
+### 기능 목록
+- [x] JsonUtils
+  - [x] Java Object를 Json 문자열로 변환한다  
+- [x] JsonView 구현
+  - [x] 응답을 Json 포맷으로 한다
+    - [x] 응답 헤더에 ContentType을 'application/json'으로 설정한다  
+  - [x] 응답 데이터의 요소가 1개인 경우 
+  - [x] 응답 데이터의 요소가 2개 이상인 경우 
+- [x] RequestParameterUtils 
+  - [x] 클라이언트의 요청 파라미터를 QeuryString 혹은 RequestBody에서 구한다
+- [x] UserApiController 구현
+  - [x] Annotation 기반 Controller로 구현한다
+  - [x] 회원가입 `POST /api/users`
+  - [x] 회원조회 `GET /api/users`
+  - [x] 회원수정 `PUT /api/users`
```
미션을 진행하기 전에 먼저 요구사항을 정리해주셨네요. 👍

yongdae commented on src/test/java/core/mvc/JsonViewTest.java:
```diff
@@ -52,11 +52,12 @@ void render_over_two_element() throws Exception {
         Car expected = new Car("Black", "Sonata");
         model.put("car", expected);
-        model.put("name", "포비");
+        model.put("name", "pobi");
```
<img width="898" alt="스크린샷 2022-08-16 오후 1 26 02" src="https://user-images.githubusercontent.com/4317804/184797664-9a569932-8962-4401-9643-8b89115025a4.png">

한글 데이터도 테스트가 통과 될 수 있도록 이 힌트를 적용해보면 어떨까요? 🤔 

<img width="796" alt="스크린샷 2022-08-16 오후 1 25 28" src="https://user-images.githubusercontent.com/4317804/184797602-eff56ca0-40e2-4faa-9e6a-8b04ab8c4d5b.png">

testrace replied:
APPLICATION_JSON_UTF8_VALUE 를 사용하고 싶었는데, Deprecated 되었더라구요!

JsonView의 역할에서는 ContentType을 'application/json' 으로 응답하는게 조금 더 중요하다 생각했습니다!
애플리케이션 전체를 보면 `CharacterEncodingFilter`가 존재하고 있어서 JsonView는 CharacterEncoding보다 ContentType에
집중하는게 좋지 않을까 생각했습니다 😃 

JsonView가 한글도 처리할 수 있는 것은 저도 좋다고 생각해서 APPLICATION_JSON_UTF8_VALUE 를 직접 정의해서
(스프링꺼는 deprecated.. 😢 ) 한글 데이터도 통과할 수 있도록 변경해 보겠습니다!

yongdae commented on src/test/java/core/mvc/JsonUtilsTest.java:
```diff
@@ -1,16 +1,50 @@
+        final String actual = JsonUtils.getParameter(jsonString, "name");
+
+        assertThat(actual).isNull();
+    }
```
다양하게 테스트 케이스를 정의해서 만들어주셨네요. 👍 
JsonUtils를 믿고 사용 할 수 있겠어요!

yongdae commented on src/main/java/core/mvc/JsonUtils.java:
```diff
@@ -1,21 +1,50 @@
+        }
+    }
+
+    public static String getParameter(final String jsonString, final String jsonKey) throws ObjectMapperException {
```
[gson 라이브러리](https://javadoc.io/doc/com.google.code.gson/gson/latest/com.google.gson/com/google/gson/JsonElement.html)처럼 getAsString() 같은 이름을 사용해보면 어떨까요?

testrace replied:
메서드명을 조금 더 구체적으로 명시하는게 API 사용자에게 더 도움이 될 것 같네요!
(네이밍은 어렵네요 😢 )

yongdae commented on src/main/java/core/mvc/JsonView.java:
```diff
@@ -1,12 +1,39 @@
+    }
 
+    private void write(final String jsonString, final HttpServletResponse response) {
+        try (final PrintWriter writer = response.getWriter()) {
```
Try-with-resource를 활용해주셨네요. 👍

yongdae commented on src/main/java/core/mvc/tobe/support/ModelArgumentResolver.java:
```diff
@@ -18,7 +18,8 @@
 
 public class ModelArgumentResolver implements ArgumentResolver {
```
이야기하셨던 `@RequestBody`를 미션에 적용시켜보면 어떨까요?
Json을 처리 할 수 있는 ArgumentResolver를 추가하는 것도 미션의 일부라고 생각해본다면 좋을 것 같아요. :)

testrace replied:
지난 미션(Controller 메소드 인자 맵핑)의 연장이라 생각하니 재밌을 것 같네요! 구현해 보겠습니다 😃

yongdae commented on src/main/java/next/controller/UserApiController.java:
```diff
@@ -0,0 +1,47 @@
+import next.model.User;
+
+@Controller
+public class UserApiController {
```
요구 사항에 대응하는 컨트롤러를 구현해주셨네요. 💯

yongdae commented on src/test/java/core/mvc/tobe/support/RequestParameterUtilsTest.java:
```diff
@@ -0,0 +1,83 @@
+import org.junit.jupiter.api.Test;
+import org.springframework.mock.web.MockHttpServletRequest;
+
+class RequestParameterUtilsTest {
```
👍

yongdae reviewed (changes requested):
안녕하세요. 용주님!
라이브러리 구현 미션 첫단계를 잘 구현해주셨네요. 💯 

소소하지만 진행해주신 부분에 대해 몇 가지 의견을 남겨 놓았습니다.
다음 단계로 넘어가기보다는 이 내용은 여기서 마무리 짓고 가는게 좋을 것 같아
Request changes로 진행하겠습니다. 😄

testrace committed:
docs: 1단계 피드백 정리

testrace committed:
fix: JsonView ContentType 변경

MediaType.APPLICATION_JSON_VALUE 에서
APPLICATION_JSON_UTF8_VALUE 으로 변경함
스프링의 MediaType.APPLICATION_JSON_UTF8_VALUE 이 Deprecated 되어 있어
직접 상수를 정의하여 사용

testrace committed:
refactor: JsonUtils 메서드명 수정

자바스크립트의 JSON 함수 명을 참고
toObject -> parse
toJson -> stringify
getParameter -> getAsStringOrNull
getParameter 는 gson 라이브러리의 메서드명을 참고하였으며
파라미터라는 표현보다 속성의 값을 String 타입으로 리턴한다는 것을 명시할 수 있음.
null 이 리턴될 수 있어서 메서드명 뒤에 OrNull 을 추가하였음.

testrace committed:
feat: RequestBody 애노테이션 추가

Http 요청의 PostData(RequestBody) 를 읽기 위한 애너테이션

testrace committed:
feat: RequestBodyQueryStringArgumentResolver

RequestBody 애노테이션 적용이 되어 있어야 파라미터를 변환할 수 있다

testrace committed:
refactor: ParameterizedTest로 변경

동일한 테스트 코드의 true/false 검증이기에 ParameterizedTest로 변경하여
테스트의 의도와 가독성 향상 목적

testrace committed:
refactor: ModelArgumentResolver 원복

testrace committed:
refactor: RequestBodyArgumentResolver 클래스명 변경

RequestBodyQueryStringArgumentResolver -> RequestBodyArgumentResolver
RequestParameterUtils를 활용하면 querystring, json 모두 지원 가능

testrace committed:
refactor: AbstractModelArgumentResolver 추상화

RequestBodyArgumentResolver에서 동일한 코드를 재사용하기 위해
request에서 파라미터를 가져오는 로직을 추상 메서드로 변경

testrace committed:
refactor: RequestBodyArgumentResolver 상속 변경

AbstractModelArgumentResolver을 상속하도록 변경함
AbstractModelArgumentResolver에서 제공하는
resolveArgument 메서드를 재사용하기 위한 목적

testrace committed:
feat: RequestBodyArgumentResolver 파라미터 변환

RequestParameterUtils를 활용한 query string, json 지원
상위 클래스인 AbstractModelArgumentResolver의 기능을 재사용

testrace committed:
feat: RequestBodyArgumentResolver 파라미터 변환 - 원시 타입

String(원시 타입으로 취급), int, long 과 같은 원시 타입 지원

testrace committed:
feat: RequestParameterUtils ThreadLocal 추가

HttpServletRequest에서 한 번 조회한 post data 를
여러 파라미터에도 적용할 수 있도록 ThreadLocal 사용

testrace committed:
refactor: RequestBodyArgumentResolver 적용

testrace commented:
안녕하세요 리뷰어님 😃 
피드백 주신 부분들을 적용해 봤습니다!

HttpServletRequest에서 PostData를 읽는 것 까지는 괜찮았는데, 두 번은 읽을 수가 없어서 파라미터가 여러개 인 경우 재활용하기 위해 
ThreadLocal을 활용했는데요! ThreadLocal에서 가장 중요한 remove() 메서드를 호출하는 시점을 ModelAndView#render 호출 이후에
호출하도록 했습니다. 
그러다 보니 ThreadLocal을 가진 객체의 위치가 좀 애매해졌는데요.. DispatcherServlet이 support 패키지에 의존하는 상황이 되었습니다.

ThreadLocal을 remove하는 시점과 ThreadLocal을 가진 객체를 어느 곳에서 관리는 하는 것이 좋을지 조금 애매하더라구요!
리뷰어님의 의견이 궁금합니다!

많은 피드백 부탁드립니다 (__)

testrace commented:
혹시 피드백 의도와 다르게 반영되었다면 힌트와 함께 RequestChange 부탁드립니다 😃

yongdae commented on src/main/java/core/mvc/tobe/support/RequestBodyArgumentResolver.java:
```diff
@@ -0,0 +1,39 @@
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class RequestBodyArgumentResolver extends AbstractModelArgumentResolver {
```
ThreadLocal을 사용함으로써 생기는 부수 효과는 예측하기 힘든 문제를 발생시킨다고 생각해요.

AbstractModelArgumentResolver를 상속받지 않고 RequestBody에 대한 ArgumentResolver를 구현한다면
ThreadLocal을 사용하지 않아도 되지 않을까요? 🤔 

ThreadLocal을 활용해주신 건 멋지다고 생각합니다!

testrace replied:
스프링에서 RequestBody 애너테이션을 하나의 메서드에서도 여러 파라미터에 지원이 가능하도록 하고 싶었습니다!
ThreadLocal에 대한 이해도가 많이 부족하여 부수 효과에 대해서는 인지를 못한 것 같아요 😢 
연습 차원에서 사용해 봤는데 테스트 코드에서는 통과를 본 것에 의미를 두고 더 공부가 필요할 것 같습니다.
피드백 감사드립니다 😄

yongdae commented on src/test/java/core/mvc/tobe/support/RequestBodyArgumentResolverTest.java:
```diff
@@ -0,0 +1,120 @@
+import org.springframework.mock.web.MockHttpServletRequest;
+import org.springframework.mock.web.MockHttpServletResponse;
+
+class RequestBodyArgumentResolverTest {
```
이번에도 RequestBodyArgumentResolver 검증을 꼼꼼하게 해주셨네요. 💯

yongdae reviewed (approved):
ThreadLocal을 사용해서 `@RequestBody`를 구현해주셨네요. 👍 
피드백을 열심히 그리고 잘 반영해주셔서 감사합니다~

그럼 다음 단계로 넘어가도록 하겠습니다.
의견 주신 부분에 대해서 답변을 남겨 놓았으니 확인 부탁드릴께요.

수고하셨습니다. 🙇
